evap/evaluation/models.py

```diff
@@ -606,13 +608,13 @@ def days_until_evaluation(self):
         return days_left
 
     def is_user_editor_or_delegate(self, user):
-        return self.contributions.filter(Q(contributor=user) | Q(contributor__in=user.represented_users.all()), can_edit=True).exists()
+        return self.contributions.filter(Q(contributor=user) | Q(contributor__in=user.represented_users.all()), can_edit=True).exists() or self.course.responsibles.filter(pk=user.pk).exists()
 
     def is_user_contributor_or_delegate(self, user):
         # early out that saves database hits since is_contributor_or_delegate is a cached_property
         if not user.is_contributor_or_delegate:
             return False
-        return self.contributions.filter(Q(contributor=user) | Q(contributor__in=user.represented_users.all())).exists()
+        return self.contributions.filter(Q(contributor=user) | Q(contributor__in=user.represented_users.all())).exists() or self.course.responsibles.filter(pk=user.pk).exists()
```

Comment:
This could only be slow if the user is a responsible or a contributor or a delegate at all, otherwise, the early-out would be taken.

This method is called if:
- Course is private or user is external and we want to determine if the user can see the evaluation (`can_user_see_evaluation`)
- Course is published but can not publish rating results or semester results are archived (`can_user_see_results_page`)
- Evaluation preview permissions check (`evaluation_preview`)
- In the evaluation detail view

For the first two, I could only find one scenario that could be slow:
In the second case, we might have multiple executions on the results_index.html template, since for each evaluation this is checked. It could be bad if we had a lot of published courses where `can_publish_rating_results` is false or results are archived for this semester - usually, that's not the case, though, right?. Still, this could be optimized by selecting the evaluations that should be shown in the view. We might want to test this.

The last two cases are just one call (one user, one evaluation). I think the query itself is rather simple, so one execution should not take that long.

Reply:
there was one more thing missing: this method (which in the meantime is called `is_user_responsible_or_contributor_or_delegate`) should also return `True` if the user is a delegate of one of the responsibles for this course even if the responsible isn't a contributor. please have a look at the implementation in the latest commit.